deepreg/model/layer.py

```diff
@@ -665,3 +666,123 @@ def call(self, inputs, training=None, **kwargs):
         r2 = self._conv3d_block(inputs=h0, training=training)
         h1 = self._residual_block(inputs=[r2, r1], training=training)
         return h1
+
+
+class BSplines3DTransform(tf.keras.layers.Layer):
+    """
+     Layer for BSplines interpolation with precomputed cubic spline filters.
+     It assumes a full sized image from which: (1) it compute the contol points values by downsampling the initial
+     image (2) performs the interpolation and (3) crops the image around the valid values.
+
+    :param cp_spacing: _int_ or tuple of three _ints_ specifying the spacing (in pixels) in each dimension.
+                      When a single _int_ is used, the same spacing to all dimensions is used
+    :param kwargs:
+    """
+
+    def __init__(self, cp_spacing: (int, tuple), **kwargs):
+
+        super().__init__(**kwargs)
+
+        if isinstance(cp_spacing, int):
+            self.cp_spacing = (cp_spacing, cp_spacing, cp_spacing)
+        else:
+            self.cp_spacing = cp_spacing
+        self.filters = []
+
+    def build(self, input_shape):

```

Comment:
if you don't mind, put `input_shape:tuple` in the input arg, so that IDE can help people to identify potential bugs